src/signals/settings/users/Detail/components/UserForm/index.js

```diff
@@ -40,75 +40,141 @@ const statusOptions = [
 
 const DEFAULT_STATUS_OPTION = 'true';
 
-const UserForm = ({ data, onCancel, onSubmitForm, readOnly }) => {
+const reducer = (state, { field, value }) => ({ ...state, [field]: value });
+
+const UserForm = ({ data, onCancel, onSubmit, readOnly }) => {
   const departments = useSelector(makeSelectDepartments);
-  const departmentList = departments.list.map(({ id, name }) => ({ id, value: name }));
+  const departmentList = departments.list.map(({ id, name }) => ({ id: String(id), value: name }));
+
   const userDepartments = data?.profile?.departments
-    .map(department => departmentList.find(({ value }) => value === department))
+    ?.map(department => departmentList.find(({ value }) => value === department))
     .filter(Boolean) || [];
 
+  const [state, dispatch] = React.useReducer(reducer, {
+    username: data.username,
+    first_name: data.first_name,
+    last_name: data.last_name,
+    note: data.profile && data.profile.note,
+    is_active: data.is_active === undefined ? DEFAULT_STATUS_OPTION : `${data.is_active}`,
+    departments: userDepartments,
+  });
+
+  const onChangeEvent = useCallback(event => {
+    onChange(event.target.name, event.target.value);
+  }, [onChange]);
+
+  const onChange = useCallback((field, value) => { dispatch({ field, value }); }, [dispatch]);
+
+  const getFormData = useCallback(() => {
+    const form = { ...data, profile: { ...data.profile } };
+
+    form.username = state.username;
+    form.first_name = state.first_name;
+    form.last_name = state.last_name;
+    form.is_active = state.is_active === 'true';
+    form.profile.note = state.note;
+    form.profile.departments = state.departments.map(department => department.value);;
+
+    const postPatch = { ...form, profile: { ...form.profile } };
+
+    delete postPatch.profile.departments;
+
+    postPatch.profile.department_ids = Object
+      .values(state.departments)
+      .map(department => department.id);
+
+    return { form, postPatch };
+  }, [data, state]);
+
+  const onSubmitForm = useCallback(
+    event => {
+      event.preventDefault();
+      onSubmit(getFormData(event));
```

Comment:
`event` Doesn't seem to be used by `getFormData`.

Reply:
Sharp, done!